polaris-core/src/main/java/org/apache/polaris/core/persistence/BasePersistence.java

```diff
@@ -0,0 +1,366 @@
+   *
+   * @param callCtx call context
+   * @param entity entity to delete
+   */
+  void deleteEntity(@Nonnull PolarisCallContext callCtx, @Nonnull PolarisBaseEntity entity);
+
+  /**
+   * Delete the specified grantRecord to the grant_records table.
+   *
+   * @param callCtx call context
+   * @param grantRec entity record to delete.
+   */
+  void deleteFromGrantRecords(
+      @Nonnull PolarisCallContext callCtx, @Nonnull PolarisGrantRecord grantRec);
+
+  /**
+   * Delete the all grant records in the grant_records table for the specified entity. This method
+   * will delete all grant records on that securable entity and also all grants to that grantee
+   * entity assuming that the entity is a grantee (catalog role, principal role or principal).
+   *
+   * @param callCtx call context
+   * @param entity entity whose grant records to and from should be deleted
+   * @param grantsOnGrantee all grants to that grantee entity. Empty list if that entity is not a
+   *     grantee
+   * @param grantsOnSecurable all grants on that securable entity
+   */
+  void deleteAllEntityGrantRecords(
+      @Nonnull PolarisCallContext callCtx,
+      @Nonnull PolarisEntityCore entity,
+      @Nonnull List<PolarisGrantRecord> grantsOnGrantee,
+      @Nonnull List<PolarisGrantRecord> grantsOnSecurable);
+
+  /**
+   * Delete Polaris entity and grant record metadata from all tables within the realm defined by the
+   * contents of the {@code callCtx}
+   *
+   * @param callCtx call context
+   */
+  void deleteAll(@Nonnull PolarisCallContext callCtx);
+
+  /**
+   * Lookup an entity given its catalog id (which can be {@link
+   * org.apache.polaris.core.entity.PolarisEntityConstants#NULL_ID} for top-level entities) and its
+   * entityId.
+   *
+   * @param callCtx call context
+   * @param catalogId catalog id or NULL_ID
+   * @param entityId entity id
+   * @return null if the entity was not found, else the retrieved entity.
+   */
+  @Nullable
+  PolarisBaseEntity lookupEntity(
+      @Nonnull PolarisCallContext callCtx, long catalogId, long entityId);
+
+  /**
+   * Lookup an entity given its catalogId, parentId, typeCode, and name.
+   *
+   * @param callCtx call context
+   * @param catalogId catalog id or {@link
+   *     org.apache.polaris.core.entity.PolarisEntityConstants#NULL_ID} for top-level entities like
+   *     CATALOG, PRINCIPAL and PRINCIPAL_ROLE. Note that by convention, a catalog itself has
+   *     NULL_ID for its catalogId since the catalog is not "nested" under itself or any other
+   *     catalog.
+   * @param parentId id of the parent
+   * @param typeCode the PolarisEntityType code of the entity to lookup
+   * @param name the name of the entity
+   * @return null if the specified entity does not exist
+   */
+  @Nullable
+  PolarisBaseEntity lookupEntityByName(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      int typeCode,
+      @Nonnull String name);
```

Comment:
The (most dominant) caller of this is when paths within a catalog are resolved.

We already have all "exploded paths" available at the call size, so I guess it's more ergonomic to delegate the resolution of these "exploded paths" to something that specialized for that. The nice side effect would be that call sites wouldn't have to deal with parent-child relationships but only with paths.